lib/blockchain/bloom.ex

```diff
@@ -0,0 +1,59 @@
+defmodule Blockchain.Bloom do
+  use Bitwise
+
+  @spec create(binary()) :: integer()
+  def create(data) when is_binary(data) do
+    bloom(0, data)
+  end
+
+  @spec add(integer(), binary()) :: integer()
+  def add(bloom_number, data) when is_binary(data) do
+    bloom(bloom_number, data)
+  end
+
+  @spec contains?(integer(), binary()) :: boolean()
+  def contains?(current_bloom, val)
+      when is_integer(current_bloom) and
+           is_binary(val) do
+    bloom = create(val)
+
+    (bloom &&& current_bloom) == bloom
+  end
+
+  @spec bloom(integer(), binary()) :: integer()
+  defp bloom(number, data) do
+    bits =
+      data
+      |> sha3_hash
+      |> bit_numbers
+
+    number |> add_bits(bits)
+  end
+
+  @spec sha3_hash(binary()) :: binary()
+  defp sha3_hash(data) do
+    data |> :keccakf1600.sha3_256
+  end
+
+  @spec add_bits(integer(), [integer()]) :: integer()
+  defp add_bits(bloom_number, bits) do
+    bits
+    |> Enum.reduce(bloom_number, fn(bit_number, bloom) ->
+      bloom ||| (1 <<< bit_number)
+    end)
+  end
+
+  @spec bit_numbers(binary()) :: [integer()]
```

Comment:
Everything looks great here, can you explain this function since I'm having a hard time wrapping my head around it.

Reply:
@hayesgm no problem. bit_numbers/1 is the implementation of (26)-(27) equations of the yellow paper
![yellowe-paper](https://i.imgur.com/aCkPNW1.png)

it takes three the low order 11-bits of the first three double-bytes of the SHA3 hash
For example, let's take sha3 hash of the "rock" binary
```elixir
iex(1)> "rock" |> :keccakf1600.sha3_256
<<16, 151, 126, 77, 104, 16, 141, 65, 132, 8, 188, 147, 16, 182, 15, 198, 208,
  167, 80, 198, 60, 206, 244, 44, 251, 14, 173, 35, 171, 115, 209, 2>>
```
The first three double-bytes are {16, 151}, {126, 77}, {104, 16}
Let's consider
> ((a1 <<< 8) + a2) &&& 2047

```elixir
iex(2)> 16 |> Integer.to_string(2) // binary representation of the first element of the first pair
"10000"
iex(3)> 151 |> Integer.to_string(2) //  binary representation of the second element of the first pair
"10010111"
iex(4)> (16 <<< 8) |> Integer.to_string(2) // the <<< operator does left bitshift to set high order bits of the first pair
"1000000000000"
iex(5)> ((16 <<< 8) + 151) |> Integer.to_string(2) // binary representation of the first double byte pair
"1000010010111" // or 4247 in decimal representation
iex(6)> ((16 <<< 8) + 151) &&& 2047 // then we calculate modulo (because our Bloom filter has only 256 bytes - 2048 bits)
151
```

Actually I wrote a [blog post](http://www.badykov.com/ethereum/2017/10/29/ethereum-bloom-filter/) about bloom filters in ethereum

Reply:
Can't we use binary matching here to make this easier?

```
iex(1)> <<first_part::integer-size(16), rest::bitstring>> = "rock" |> :keccakf1600.sha3_256
<<16, 151, 126, 77, 104, 16, 141, 65, 132, 8, 188, 147, 16, 182, 15, 198, 208,
  167, 80, 198, 60, 206, 244, 44, 251, 14, 173, 35, 171, 115, 209, 2>>
iex(2)> first_part
4247
iex(3)> rest
<<126, 77, 104, 16, 141, 65, 132, 8, 188, 147, 16, 182, 15, 198, 208, 167, 80,
  198, 60, 206, 244, 44, 251, 14, 173, 35, 171, 115, 209, 2>>
```

Reply:
@hayesgm good point! thanks